syntax = "proto3";

package example;

// Define a message for VectorParam
message VectorParam {
  string name = 1;
  string opcode = 2;
  repeated int32 input = 3;
  repeated int32 other = 4;
}

// message for GEMM, pooling, and layer norm
message MatrixParam {
  string name = 1;
  string opcode = 2;
  repeated int32 input = 3;
  repeated int32 weight = 4;
  repeated int32 bias = 5;
  repeated int32 stride = 6;
  repeated int32 padding = 7;
  repeated int32 dilation = 8;
  bool transposed = 9;
  bool ceil_mode = 10;
}

// message for reduction and softmax
message ReduceParam {
  string name = 1;
  string opcode = 2;
  repeated int32 input = 3;
  repeated int32 dim = 4;
  bool keepdim = 5;
}

// message for transpose and permute
message ShapeParam {
  string name = 1;
  string opcode = 2;
  repeated int32 input = 3;
  repeated int32 dims = 4;
}

// One of GEMM, reduce, or shape change
message Param {
  oneof param_type {
    MatrixParam matrix_param = 1;
    ReduceParam reduce_param = 2;
    ShapeParam shape_param = 3;
  }
  repeated VectorParam vector_params = 4;
  bool fused = 5;
}

message ModelParams {
    repeated Param params = 1;
}
